### ISSUE TITLE
some question of pad_with_border

### ISSUE THREAD

**Issue by Nickkk1124** [yongxuUSTC/sednn#8]
Hello:
Really impressed by your work and got a few questions in terms of how you process the data.

![31131456_1863410970346818_1790379927908909056_n](https://user-images.githubusercontent.com/36222959/39170816-6f889a30-47cf-11e8-9b23-8f3ca4c84ace.jpg)

Do pad_with_border mean this?

Many thanks,
Nick

**Comment by Nickkk1124** [comment]
Sorry
In addition, I would like to ask if I want to use this speech-enhanced system on the front of the asr. How do I do this?

Many thanks,
Nick

**Comment by qiuqiangkong** [comment]
Hi Nick,


    The picture you show is correct. pad_with_border simply extend the left and right border.


    You may obtain enhanced speech from by running this code. Then ASR may apply post-hoc.


Best wishes,

Qiuqiang

________________________________

**Comment by Nickkk1124** [comment]
Hello Qiuqiang,


    Mat_2d_to_3d is to convert features to (n_segs, n_concat, n_freq).


    The center frame of the first round of stacking frames is t=1, and the center frame of the second round of stacking frames should not be t=2, but as shown in the following figure, why is the center frame of the second round of stacking frames t=4?

![31179984_1863648643656384_4102975115338186752_n](https://user-images.githubusercontent.com/36222959/39188624-c9a5b25e-4802-11e8-99b8-bdab94c7b24e.jpg)

Many thanks,


Nick

**Comment by yongxuUSTC** [comment]
Hi Nick,

Yes, you can use the enhanced features for ASR. But maybe you should use retraining or joint-training of your backend acoustic model for ASR.

Good luck.

Best regards,
yong


--------------------------------------------------------
Dr. Yong XU
https://sites.google.com/view/xuyong/home
 
From: Nickkk1124
Date: 2018-04-24 09:57
To: yongxuUSTC/sednn
CC: Subscribed
Subject: Re: [yongxuUSTC/sednn] some question of pad_with_border (#8)
Sorry
In addition, I would like to ask if I want to use this speech-enhanced system on the front of the asr. How do I do this?
Many thanks,
Nick

**Comment by qiuqiangkong** [comment]
Hi Nick,


    In the picture you draw, it is correct. center frame=1 and center frame=4 in your drawing. It also depends on the hop.


"The "enhanced features for ASR" you mentioned, do you mean the magnitudes of log power spectrogram?"


- It means either enhanced spectrogram or log power spectrogram.


"Do you think using recover enhanced wav as ASR input is feasible?"

It is feasible if the dataset is small. However bare in mind any speech denoising


- method will lose some information. Some work did a joint enhancement and recognition.


"What would you recommend about applying the enhancement system to dealing with the environmental noise?"


- I think applying on environmental noise should be fine, as long as the noise for training covers most environmental noise.


Best wishes,


Qiuqiang

________________________________

**Comment by akshayaCap** [comment]
Hello Qiuqiang,

This is a great work. It will be of great help if you could elaborate on below points mentioned by you in above discussion.

"- method will lose some information. Some work did a joint enhancement and recognition."

I get the point of information loss. Can you please tell more about Joint enhancement and recognition?

Is it like two 2 DNN models interlinked or preprocessing and ASR. 

Thank-you.

**Comment by qiuqiangkong** [comment]
Hi Nick,


    If speech enhancement and ASR are done separately, the ASR performance might be reduced. Because sometimes speech enhancement will also move out some useful information of a speech. However, if they are combined to a single neural network it might be helpful. For example, use speech enhancement as lower layer of a neural network and use ASR as a high layer neural network. The loss function can combine the ASR and speech enhancement. It is just my conjecture and I am not aware if there is such work or not.


Best wishes,


Qiuqiang

________________________________

**Comment by yongxuUSTC** [comment]
Hi Nick,

Yes, there are joint SE & ASR training papers: 
https://www.isca-speech.org/archive/interspeech_2014/i14_0616.html
https://ieeexplore.ieee.org/abstract/document/7178797/

Best regards,
yong


----------------------------------------------------------
Yong XU
https://sites.google.com/view/xuyong/home
 
From: qiuqiangkong
Date: 2018-07-06 03:55
To: yongxuUSTC/sednn
CC: yong xu @ seattle; Comment
Subject: Re: [yongxuUSTC/sednn] some question of pad_with_border (#8)
Hi Nick,


If speech enhancement and ASR are done separately, the ASR performance might be reduced. Because sometimes speech enhancement will also move out some useful information of a speech. However, if they are combined to a single neural network it might be helpful. For example, use speech enhancement as lower layer of a neural network and use ASR as a high layer neural network. The loss function can combine the ASR and speech enhancement. It is just my conjecture and I am not aware if there is such work or not.


Best wishes,


Qiuqiang

________________________________
